### DISCUSSION TITLE
uhubctl Synology NAS

### DISCUSSION THREAD

**Discussion by SilverCH** [mvp/uhubctl#517] [Q&A] [answered]
Hello, has anyone installed uhubctl on a Synology NAS or can they explain how to run it within the NAS?

**Comment by mvp** [comment]
I don't know much about Synology except that it is using some Linux distro. With that in mind, you should be able to compile uhubctl from source, as long as you get following packages installed on Synology: `gcc`, `make`, `libusb`, `git`, `pkg-config` - perhaps you need to find out some kind of Synology web interface which allows you to install additional Linux packages. Once these packages are present and you have ssh access to your Synology, you just need to run:
```
git clone https://github.com/mvp/uhubctl
make
sudo make install
```
and it all should just work, as long as you have compatible USB hub(s) attached to your Synology.

**Reply by mvp** [reply]
@SilverCH: for what it's worth, someone has apparently just recently added uhubctl to Synology package repository https://github.com/SynoCommunity/spksrc/pull/5833

**Comment by SilverCH** [comment]
I'm currently facing an issue with a Tplink UH700 USB 3.0 Hub on my Synology system. Despite the hub being advertised as compatible, I'm encountering a "not compatible" message when trying to use it.

When I run the sudo uhubctl command, I receive the following output:

`Current status for hub 1-1.2.1 [da0b:1154 Generic 4-Port USB 2.0 Hub, USB 10.02, 4 ports, ppps]
  Port 1: 0305 power 5gbps Recovery connect [9b05:7000 Iomega eGo USB 010000000000467D]
  Port 2: 0305 power 5gbps Recovery connect [5810:2008 Western Digital My Passport 0820 575838314142333034343535]
  Port 3: 0001 off connect []
  Port 4: 0001 off connect []
...
`
I tried using the sudo uhubctl -l 1-1.2.1 -p 1 -a off command to disable a port, but I received the message "No compatible devices detected at location 1-1.2.1!"

Could anyone please help me troubleshoot this compatibility issue? I would greatly appreciate any insights or suggestions you might have.

Thank you in advance for your time and assistance!

Best regards,
Nedi

**Reply by mvp** [reply]
USB 10.02 seems suspicious, almost as if bytes were swapped - it should be 2.10. Also, empty brackets `[]` is not normal.

This could be one of:
* uhubctl version is very old. You want 2.4.0 or higher (latest version is 2.5.0)
* uhubctl was linked with very old version of libusb. Ideally, it should be libusb 1.0.26+
* swapped bytes in USB version may mean wrong endianness. uhubctl was mostly tested on little endian hardware (x64, arm32/arm64). Perhaps Synology uses big endian hardware, and uhubctl has endianness bug?
* or, it could be some other bug. Unfortunately, it is rather difficult to debug this without having shell access to real hardware.

**Reply by mvp** [reply, marked as answer]
@SilverCH, I have released uhubctl 2.6.0 which should fix big-endian bug that apparently was present on Synology.
